File: packages/contracts/python/butterflylens/contracts/live_worker.py
"""Persistent live-worker wire declarations."""

from __future__ import annotations

from typing import Literal, TypedDict


WORKER_IDENTITY_SCHEMA_VERSION = "butterflylens-worker-identity:v1.0.0"
WORKER_HEARTBEAT_SCHEMA_VERSION = "butterflylens-worker-heartbeat:v1.0.0"
WORKER_LEASE_SCHEMA_VERSION = "butterflylens-worker-lease:v1.0.0"
WORKER_COMMAND_SCHEMA_VERSION = "butterflylens-worker-command:v1.0.0"
WORKER_EVENT_SCHEMA_VERSION = "butterflylens-worker-event:v1.0.0"

WORKER_COMMAND_KINDS = (
    "start_run",
    "pause_run",
    "resume_run",
    "cancel_run",
    "health_check",
    "graceful_shutdown",
)
WORKER_EVENT_KINDS = (
    "stage_started",
    "progress",
    "checkpoint_committed",
    "artifact_committed",
    "stage_succeeded",
    "stage_failed",
    "run_paused",
    "run_cancelled",
    "shutdown_started",
    "shutdown_complete",
)


class WorkerIdentity(TypedDict):
    schema_version: Literal["butterflylens-worker-identity:v1.0.0"]
    worker_id: str
    registered_at: str
    machine_profile: dict[str, object]
    capabilities: dict[str, object]
    configured_models: list[dict[str, object]]
    identity_fingerprint: str
    scientific_claim_allowed: Literal[False]


class WorkerHeartbeat(TypedDict):
    schema_version: Literal["butterflylens-worker-heartbeat:v1.0.0"]
    heartbeat_id: str
    worker_id: str
    sequence: int
    observed_at: str
    state: str
    project_id: str | None
    run_id: str | None
    lease_id: str | None
    lease_revision: int | None
    lease_expires_at: str | None
    current_stage_id: str | None
    resources: dict[str, int | None]
    queues: list[dict[str, int | str]]
    models: list[dict[str, object]]
    cache: dict[str, int]
    last_committed_artifact_fingerprint: str | None
    last_committed_at: str | None
    health_checks: list[dict[str, str]]
    heartbeat_fingerprint: str
    scientific_claim_allowed: Literal[False]


class WorkerLease(TypedDict):
    schema_version: Literal["butterflylens-worker-lease:v1.0.0"]
    lease_id: str
    project_id: str
    run_id: str
    stage_id: str
    worker_id: str
    status: str
    revision: int
    fencing_token: str
    idempotency_key: str
    issued_at: str
    acquired_at: str | None
    renewed_at: str | None
    expires_at: str
    released_at: str | None
    checkpoint_fingerprint: str | None
    cancellation_requested: bool
    cancellation_requested_at: str | None
    lease_fingerprint: str


class WorkerCommand(TypedDict):
    schema_version: Literal["butterflylens-worker-command:v1.0.0"]
    command_id: str
    worker_id: str
    project_id: str | None
    run_id: str | None
    lease_id: str | None
    expected_lease_revision: int | None
    kind: str
    idempotency_key: str
    issued_at: str
    expires_at: str
    requested_by: str
    payload: dict[str, object]
    command_fingerprint: str


class WorkerEvent(TypedDict):
    schema_version: Literal["butterflylens-worker-event:v1.0.0"]
    event_id: str
    worker_id: str
    sequence: int
    project_id: str
    run_id: str
    lease_id: str
    lease_revision: int
    kind: str
    stage_id: str
    occurred_at: str
    records_processed: int
    bytes_processed: int
    checkpoint_fingerprint: str | None
    artifact_fingerprint: str | None
    error: dict[str, object] | None
    event_fingerprint: str
    scientific_claim_allowed: Literal[False]

File: packages/contracts/src/live-worker.ts
export const WORKER_IDENTITY_SCHEMA_VERSION =
  'butterflylens-worker-identity:v1.0.0' as const
export const WORKER_HEARTBEAT_SCHEMA_VERSION =
  'butterflylens-worker-heartbeat:v1.0.0' as const
export const WORKER_LEASE_SCHEMA_VERSION =
  'butterflylens-worker-lease:v1.0.0' as const
export const WORKER_COMMAND_SCHEMA_VERSION =
  'butterflylens-worker-command:v1.0.0' as const
export const WORKER_EVENT_SCHEMA_VERSION =
  'butterflylens-worker-event:v1.0.0' as const

export const WORKER_COMMAND_KINDS = [
  'start_run',
  'pause_run',
  'resume_run',
  'cancel_run',
  'health_check',
  'graceful_shutdown',
] as const

export const WORKER_EVENT_KINDS = [
  'stage_started',
  'progress',
  'checkpoint_committed',
  'artifact_committed',
  'stage_succeeded',
  'stage_failed',
  'run_paused',
  'run_cancelled',
  'shutdown_started',
  'shutdown_complete',
] as const

export interface WorkerIdentity {
  readonly schema_version: typeof WORKER_IDENTITY_SCHEMA_VERSION
  readonly worker_id: string
  readonly registered_at: string
  readonly machine_profile: {
    readonly platform: 'macos' | 'linux'
    readonly architecture: 'arm64' | 'x86_64'
    readonly os_version: string
    readonly chip_label: string
    readonly cpu_core_count: number
    readonly unified_memory_bytes: number
    readonly mps_available: boolean
    readonly mps_runtime: string | null
  }
  readonly capabilities: {
    readonly supported_stage_ids: readonly string[]
    readonly max_queue_records: number
    readonly max_queue_bytes: number
    readonly rolling_prefetch_batches: number
    readonly checkpoint_format: 'parquet-manifest-v1'
    readonly graceful_shutdown_supported: true
  }
  readonly configured_models: readonly {
    readonly role: 'yoloe_router' | 'bioclip_embedder'
    readonly model_id: string
    readonly revision: string
    readonly weights_sha256: string
    readonly preprocessing_fingerprint: string
    readonly licence_status: 'approved' | 'blocked'
    readonly device: 'mps' | 'cpu'
  }[]
  readonly identity_fingerprint: string
  readonly scientific_claim_allowed: false
}

export interface WorkerHeartbeat {
  readonly schema_version: typeof WORKER_HEARTBEAT_SCHEMA_VERSION
  readonly heartbeat_id: string
  readonly worker_id: string
  readonly sequence: number
  readonly observed_at: string
  readonly state: 'starting' | 'idle' | 'leased' | 'running' | 'paused' | 'draining' | 'degraded'
  readonly project_id: string | null
  readonly run_id: string | null
  readonly lease_id: string | null
  readonly lease_revision: number | null
  readonly lease_expires_at: string | null
  readonly current_stage_id: string | null
  readonly resources: Readonly<Record<string, number | null>>
  readonly queues: readonly Readonly<Record<string, number | string>>[]
  readonly models: readonly Readonly<Record<string, unknown>>[]
  readonly cache: Readonly<Record<string, number>>
  readonly last_committed_artifact_fingerprint: string | null
  readonly last_committed_at: string | null
  readonly health_checks: readonly Readonly<Record<string, string>>[]
  readonly heartbeat_fingerprint: string
  readonly scientific_claim_allowed: false
}

export interface WorkerLease {
  readonly schema_version: typeof WORKER_LEASE_SCHEMA_VERSION
  readonly lease_id: string
  readonly project_id: string
  readonly run_id: string
  readonly stage_id: string
  readonly worker_id: string
  readonly status: 'offered' | 'active' | 'released' | 'expired' | 'revoked'
  readonly revision: number
  readonly fencing_token: string
  readonly idempotency_key: string
  readonly issued_at: string
  readonly acquired_at: string | null
  readonly renewed_at: string | null
  readonly expires_at: string
  readonly released_at: string | null
  readonly checkpoint_fingerprint: string | null
  readonly cancellation_requested: boolean
  readonly cancellation_requested_at: string | null
  readonly lease_fingerprint: string
}

export interface WorkerCommand {
  readonly schema_version: typeof WORKER_COMMAND_SCHEMA_VERSION
  readonly command_id: string
  readonly worker_id: string
  readonly project_id: string | null
  readonly run_id: string | null
  readonly lease_id: string | null
  readonly expected_lease_revision: number | null
  readonly kind: (typeof WORKER_COMMAND_KINDS)[number]
  readonly idempotency_key: string
  readonly issued_at: string
  readonly expires_at: string
  readonly requested_by: string
  readonly payload: {
    readonly stage_id: string | null
    readonly checkpoint_fingerprint: string | null
    readonly reason: string | null
  }
  readonly command_fingerprint: string
}

export interface WorkerEvent {
  readonly schema_version: typeof WORKER_EVENT_SCHEMA_VERSION
  readonly event_id: string
  readonly worker_id: string
  readonly sequence: number
  readonly project_id: string
  readonly run_id: string
  readonly lease_id: string
  readonly lease_revision: number
  readonly kind: (typeof WORKER_EVENT_KINDS)[number]
  readonly stage_id: string
  readonly occurred_at: string
  readonly records_processed: number
  readonly bytes_processed: number
  readonly checkpoint_fingerprint: string | null
  readonly artifact_fingerprint: string | null
  readonly error: {
    readonly code: string
    readonly message: string
    readonly retryable: boolean
  } | null
  readonly event_fingerprint: string
  readonly scientific_claim_allowed: false
}
